src/openms/include/OpenMS/ANALYSIS/TOPDOWN/SpectrumDeconvolution.h

```diff
@@ -0,0 +1,159 @@
+// OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+// WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
+// OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
+// ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// --------------------------------------------------------------------------
+// $Maintainer: Kyowon Jeong, Jihyung Kim $
+// $Authors: Kyowon Jeong, Jihyung Kim $
+// --------------------------------------------------------------------------
+
+#pragma once
+
+#include <OpenMS/APPLICATIONS/TOPPBase.h>
+#include <OpenMS/CONCEPT/Types.h>
+#include <OpenMS/CHEMISTRY/ISOTOPEDISTRIBUTION/IsotopeDistribution.h>
+#include <OpenMS/CHEMISTRY/ISOTOPEDISTRIBUTION/CoarseIsotopePatternGenerator.h>
+#include <OpenMS/MATH/STATISTICS/StatisticFunctions.h>
+#include "boost/dynamic_bitset.hpp"
+#include <iostream>
+#include <iomanip>
+#include <chrono>
+
+#include <OpenMS/ANALYSIS/TOPDOWN/FLASHDeconvHelperStructs.h>
+#include <OpenMS/ANALYSIS/TOPDOWN/PeakGroupScoring.h>
+#include <OpenMS/ANALYSIS/TOPDOWN/FLASHDeconvAlgorithm.h>
+#include <OpenMS/ANALYSIS/TOPDOWN/PeakGroup.h>
+
+
+namespace OpenMS
+{
+  //This class performs spectrum level deconvolution.
+  class OPENMS_DLLAPI SpectrumDeconvolution
+  {
+  public:
+    //FLASHDeconv parameter
+    typedef FLASHDeconvHelperStructs::Parameter Parameter;
+    //Precalculated averagine
+    typedef FLASHDeconvHelperStructs::PrecalculatedAveragine PrecalculatedAveragine;
+    //Peaks with log mz transformation
+    typedef FLASHDeconvHelperStructs::LogMzPeak LogMzPeak;
+
+    /// default constructor
+    SpectrumDeconvolution(MSSpectrum &spec, Parameter &param);
+
+    /// default destructor
+    ~SpectrumDeconvolution();
+
+    bool empty();
+
+    //Main function
+    std::vector<PeakGroup> &getPeakGroupsFromSpectrum(std::vector<std::vector<Size>> &prevMassBinVector,
+                                                      std::vector<double> &prevMinBinLogMassVector,
+                                                      FLASHDeconvHelperStructs::PrecalculatedAveragine &avg,
+                                                      unsigned int msLevel);
+
+    //Stores log mz peaks
+    std::vector<LogMzPeak> logMzPeaks;
+
+  protected:
+    MSSpectrum &spec;
+    Parameter &param;
+
+    //Bins for mass and mz. Bin size is deteremine by Parameter.tolerance
+    //massBinsForThisSpectrum stores the selected bins only for this spectrum
+    boost::dynamic_bitset<> massBinsForThisSpectrum;
+    //massBins stores the selected bins for this spectrum + overlapped spectrum (previous a few spectra).
+    boost::dynamic_bitset<> massBins;
+    //mzBins stores the binned log mz peaks
+    boost::dynamic_bitset<> mzBins;
+    //peakGroups stores the decovnoluted mass peak groups
+    std::vector<PeakGroup> peakGroups;
+
+    //This stores the "universal pattern"
+    double *filter{};
```

Comment:
would be great to get rid of the pointer stuff here. std::vector<double> - you can get the ptr via http://www.cplusplus.com/reference/vector/vector/data/